src/lib.rs

```diff
@@ -308,47 +319,99 @@ pub fn path_validation(project_path: &Path) -> Result<PathBuf> {
 #[cfg(test)]
 mod tests {
     use super::*;
+    use proptest::prelude::*;
+    use proptest_derive::Arbitrary;
 
-    // Test other lib internal functions
-    #[test]
-    fn define_name_valid_test() {
-        assert!(define_name("test-project", Path::new("~/Desktop/project")).is_ok());
+    static VALID_LICENSES: [&str; 3] = ["MIT", "Apache-2.0", "GPL-3.0"];
+
+    #[derive(Debug, Arbitrary)]
+    struct LicenseTest {
+        license_str: String,
     }
-    #[test]
-    fn define_name_emptyname_test() {
-        assert!(define_name("", Path::new("~/Desktop/MyProject")).is_ok());
+
+    proptest! {
+        #[test]
+        fn define_license_proptest(data: LicenseTest) {
+
+            match define_license(&data.license_str) {
+                Err(Error::NoLicense) => prop_assert!(&data.license_str.is_empty()),
+                Err(Error::InvalidLicense) => prop_assert!(!VALID_LICENSES.contains(&&*data.license_str)),
+                Ok(_) => prop_assert!(VALID_LICENSES.contains(&&*data.license_str)),
+                _ => {},
+            }
+        }
     }
-    #[test]
-    fn emptypath_test() {
-        assert!(path_validation(Path::new("")).is_err())
+
+    fn path_strategy() -> impl Strategy<Value = PathBuf> {
+        "\\PC*".prop_map(PathBuf::from)
+    }
+    proptest! {
+        #[test]
+        fn define_name_proptest(project_name in "\\PC*", project_path in path_strategy()) {
+            let project_path_str = project_path.file_name().and_then(|x| x.to_str()).unwrap_or("");
+
+            match define_name(&project_name, &project_path) {
+                Ok(name) => {
+                    if !project_name.is_empty() && project_name.is_ascii() {
+                        prop_assert_eq!(name, &project_name)
+                    } else {
+                        prop_assert_eq!(name, project_path_str)
+                    }
+                }
+                Err(Error::UTF8Check) => prop_assert!(project_path_str.is_empty() || !project_path_str.is_ascii() || project_path_str.contains('/')),
+                _ => {},
+            }
+        }
     }
+
     #[test]
-    fn define_license_valid_test() {
-        assert!(define_license("AFL-3.0").is_ok())
+    fn test_valid_path() {
+        let valid_path = Path::new("~");
+        let result = path_validation(valid_path);
+        assert!(result.is_ok());
     }
+
     #[test]
-    fn define_license_invalid_test() {
-        assert!(define_license("POL-3.0").is_err());
+    fn test_creation_error() {
+        let unwritable_path = Path::new("Desktop/not_writable");
+        let result = path_validation(unwritable_path);
+        assert!(result.is_err());
```

Comment:
Same as above. Can we use `matches` to show the specific error returned in this case?